orca-core/src/main/java/com/netflix/spinnaker/orca/locks/ExecutionCompleteLockReleasingListener.java

```diff
@@ -0,0 +1,48 @@
+package com.netflix.spinnaker.orca.locks;
+
+import com.netflix.spinnaker.orca.events.ExecutionComplete;
+import com.netflix.spinnaker.orca.pipeline.AcquireLockStage;
+import com.netflix.spinnaker.orca.pipeline.model.Execution;
+import com.netflix.spinnaker.orca.pipeline.persistence.ExecutionRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationListener;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ExecutionCompleteLockReleasingListener implements ApplicationListener<ExecutionComplete> {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final ExecutionRepository executionRepository;
+  private final LockManager lockManager;
+  private final LockingConfigurationProperties lockingConfigurationProperties;
+
+  @Autowired
+  public ExecutionCompleteLockReleasingListener(ExecutionRepository executionRepository,
+                                                LockManager lockManager,
+                                                LockingConfigurationProperties lockingConfigurationProperties) {
+    this.executionRepository = executionRepository;
+    this.lockManager = lockManager;
+    this.lockingConfigurationProperties = lockingConfigurationProperties;
+  }
+
+  @Override
+  public void onApplicationEvent(ExecutionComplete event) {
+    if (!lockingConfigurationProperties.isEnabled()) {
+      return;
+    }
+    if (event.getStatus().isHalt()) {
+      Execution execution = executionRepository.retrieve(event.getExecutionType(), event.getExecutionId());
+      execution.getStages().forEach(s -> {
+        if (AcquireLockStage.PIPELINE_TYPE.equals(s.getType())) {
+          try {
+            LockContext lc = s.mapTo("/lock", LockContext.LockContextBuilder.class).withStage(s).build();
+            lockManager.releaseLock(lc.getLockName(), lc.getLockValue(), lc.getLockHolder());
+          } catch (LockFailureException lfe) {
+            logger.info("Failure releasing lock in ExecutionCompleteLockReleasingListener - ignoring", lfe);
```

Comment:
What's the effect of failing to release the lock? The log level & ignoring sounds like it's not a big deal, but not releasing sounds like a big deal.

Reply:
Worst case of failing to release here would be that the lock still exists and is unavailable until the TTL expires (120s by default).

I feel like this is more likely to happen if the execution got in a bad way and stalled and did not retain the lock, and releasing failed because the lock no longer exists - in which case this shouldn't matter because the execution is done